gsr_booking/booking_logic.py

```diff
@@ -0,0 +1,255 @@
+                "partial_success": False,
+                "error": "No members in the group have enabled their pennkey to be used for group bookings",
+            }
+
+        # randomize order, and put requester first in the order
+        random.shuffle(members)
+        for (i, member) in enumerate(members):  # puts the requester as the first person
+            if member["username"] == requester_pennkey:
+                temp = members[0]
+                members[0] = member
+                members[i] = temp
+                break
+
+        if is_wharton:  # huntsman reservation
+            return {
+                "complete_success": False,
+                "partial_success": False,
+                "error": "Unable to book huntsman rooms yet",
+            }
+        else:  # lib reservation
+            # Find the first timeslot to book for (next_start, next_end)
+            START_DATE = datetime.datetime.strptime(start, self.DATE_FORMAT_STR)
+            END_DATE = datetime.datetime.strptime(end, self.DATE_FORMAT_STR)
+            next_start = START_DATE
+            next_end = min(END_DATE, next_start + datetime.timedelta(hours=self.MAX_SLOT_HRS))
+
+            # loop through each member, and attempt to book on their behalf
+            bookings = {}
+            failed_members = []  # store the members w/ failed bookings in here
+
+            for member in members:
+                if next_end - next_start < datetime.timedelta(hours=0.1):
+                    break
+                if member["user__email"] is "":
+                    continue
+
+                # make request to labs-api-server
+                success, error = self.book_room_for_user(
+                    room,
+                    lid,
+                    next_start.strftime(self.DATE_FORMAT_STR),
+                    next_end.strftime(self.DATE_FORMAT_STR),
+                    member["user__email"],
+                )
+                if success:
+                    key = f"{lid}_{room}"
+                    if not key in bookings:
+                        bookings[key] = []
+                    new_bookings = self.split_booking(
+                        next_start, next_end, member["username"], True
+                    )
+                    bookings[key].extend(new_bookings)
+
+                    next_start = next_end
+                    next_end = min(
+                        END_DATE, next_start + datetime.timedelta(hours=self.MAX_SLOT_HRS)
+                    )
+                elif self.is_fatal_error(error):
+                    result_json = self.construct_bookings_json_obj(
+                        bookings, lid, room, END_DATE, next_start, next_end, error
+                    )
+                    return result_json
+                else:
+                    failed_members.append(member)
+
+            # if unbooked slots still remain and not all booking requests succeeded, loop through each member again
```

Comment:
I know we talked about this before, but looking at the code, just wondering why we don't get how much credit a user has left before attempting a booking in the first place? Seems like this might just be duplicating some work. Feel like the algorithm would be shorter if we didn't do the "blind" booking first.

Reply:
The algorithm would be shorter, but slower on average if we skipped the "blind" booking. This is because getting reservations for a user takes ~2x times longer than making a booking for that user. Also, although including "blind" booking does increase the worse case time for booking, in the average case we would only need to get reservation credits for a few users(assuming that some people in the group have not booked anything).

I talked with Josh and he suggested that in the future we can use the booking credit data stored in SQL. Although it only captures the bookings made through pennmobile, we would be able to get reservation credits almost instantly, allowing us to remove the blind booking part and drastically decease time to book for group.